File: fe/src/main/java/com/cloudera/impala/analysis/FloatLiteral.java
// Copyright 2012 Cloudera Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.cloudera.impala.analysis;

import com.cloudera.impala.catalog.ColumnType;
import com.cloudera.impala.common.AnalysisException;
import com.cloudera.impala.common.NotImplementedException;
import com.cloudera.impala.thrift.TExprNode;
import com.cloudera.impala.thrift.TExprNodeType;
import com.cloudera.impala.thrift.TFloatLiteral;
import com.google.common.base.Objects;
import com.google.common.base.Preconditions;

public class FloatLiteral extends LiteralExpr {
  private double value_;

  private void init(Double value) {
    this.value_ = value.doubleValue();
    // Figure out if this will fit in a FLOAT without loosing precision.
    float fvalue;
    fvalue = value.floatValue();
    if (fvalue == value.doubleValue()) {
      type_ = ColumnType.FLOAT;
    } else {
      type_ = ColumnType.DOUBLE;
    }
  }

  public FloatLiteral(Double value) {
    init(value);
  }

  /**
   * C'tor forcing type, e.g., due to implicit cast
   */
  public FloatLiteral(Double value, ColumnType type) {
    this.value_ = value.doubleValue();
    this.type_ = type;
  }

  public FloatLiteral(String value) throws AnalysisException {
    Double floatValue = null;
    try {
      floatValue = new Double(value);
    } catch (NumberFormatException e) {
      throw new AnalysisException("invalid floating-point literal: " + value, e);
    }
    init(floatValue);
  }

  @Override
  public String debugString() {
    return Objects.toStringHelper(this)
        .add("value", value_)
        .toString();
  }

  @Override
  public boolean equals(Object obj) {
    if (!super.equals(obj)) return false;
    return ((FloatLiteral) obj).value_ == value_;
  }

  @Override
  public String toSqlImpl() {
    return getStringValue();
  }

  @Override
  public String getStringValue() {
    return Double.toString(value_);
  }

  @Override
  protected void toThrift(TExprNode msg) {
    msg.node_type = TExprNodeType.FLOAT_LITERAL;
    msg.float_literal = new TFloatLiteral(value_);
  }

  public double getValue() {
    return value_;
  }

  @Override
  protected Expr uncheckedCastTo(ColumnType targetType) throws AnalysisException {
    Preconditions.checkState(targetType.isFloatingPointType() || targetType.isDecimal());
    if (targetType.isFloatingPointType()) {
      type_ = targetType;
    } else {
      return new CastExpr(targetType, this, true);
    }
    return this;
  }

  @Override
  public void swapSign() throws NotImplementedException {
    // swapping sign does not change the type
    value_ = -value_;
  }

  @Override
  public int compareTo(LiteralExpr o) {
    if (!(o instanceof FloatLiteral)) return -1;
    FloatLiteral other = (FloatLiteral) o;
    if (value_ > other.getValue()) return 1;
    if (value_ < other.getValue()) return -1;
    return 0;
  }
}

File: be/src/exprs/float-literal.cc
// Copyright 2012 Cloudera Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "float-literal.h"

#include <sstream>

#include "codegen/llvm-codegen.h"
#include "gen-cpp/Exprs_types.h"

using namespace llvm;
using namespace std;

namespace impala {

FloatLiteral::FloatLiteral(const ColumnType& type, void* value)
  : Expr(type) {
  DCHECK(value != NULL);
  switch (type_.type) {
    case TYPE_FLOAT:
      result_.float_val = *reinterpret_cast<float*>(value);
      break;
    case TYPE_DOUBLE:
      result_.double_val = *reinterpret_cast<double*>(value);
      break;
    default:
      DCHECK(false) << "FloatLiteral ctor: bad type: " << type_.DebugString();
  }
}

FloatLiteral::FloatLiteral(const TExprNode& node)
  : Expr(node) {
  switch (type_.type) {
    case TYPE_FLOAT:
      result_.float_val = node.float_literal.value;
      break;
    case TYPE_DOUBLE:
      result_.double_val = node.float_literal.value;
      break;
    default:
      DCHECK(false) << "FloatLiteral ctor: bad type: " << type_.DebugString();
  }
}

void* FloatLiteral::ReturnFloatValue(Expr* e, TupleRow* row) {
  FloatLiteral* l = static_cast<FloatLiteral*>(e);
  return &l->result_.float_val;
}

void* FloatLiteral::ReturnDoubleValue(Expr* e, TupleRow* row) {
  FloatLiteral* l = static_cast<FloatLiteral*>(e);
  return &l->result_.double_val;
}

Status FloatLiteral::Prepare(RuntimeState* state, const RowDescriptor& row_desc) {
  DCHECK_EQ(children_.size(), 0);
  switch (type()) {
    case TYPE_FLOAT:
      compute_fn_ = ReturnFloatValue;
      break;
    case TYPE_DOUBLE:
      compute_fn_ = ReturnDoubleValue;
      break;
    default:
      DCHECK(false) << "FloatLiteral::Prepare(): bad type: " << type_.DebugString();
  }
  return Status::OK;
}

string FloatLiteral::DebugString() const {
  stringstream out;
  out << "FloatLiteral(value=";
  switch (type()) {
    case TYPE_FLOAT:
      out << result_.float_val;
      break;
    case TYPE_DOUBLE:
      out << result_.double_val;
      break;
    default:
      DCHECK(false) << "FloatLiteral::DebugString(): bad type: " << type_.DebugString();
  }
  out << ")";
  return out.str();
}

// LLVM IR generation for float literals.  Resulting IR looks like:
//
// define double @FloatLiteral(i8** %row, i8* %state_data, i1* %is_null) {
// entry:
//   store i1 false, i1* %is_null
//   ret double 1.100000e+00
// }
Function* FloatLiteral::Codegen(LlvmCodeGen* codegen) {
  DCHECK_EQ(GetNumChildren(), 0);
  LLVMContext& context = codegen->context();
  LlvmCodeGen::LlvmBuilder builder(context);

  Function* function = CreateComputeFnPrototype(codegen, "FloatLiteral");
  BasicBlock* entry_block = BasicBlock::Create(context, "entry", function);

  Value* result = NULL;
  switch (type()) {
    case TYPE_FLOAT:
      result = ConstantFP::get(context, APFloat(result_.float_val));
      break;
    case TYPE_DOUBLE:
      result = ConstantFP::get(context, APFloat(result_.double_val));
      break;
    default:
      DCHECK(false) << "Invalid type.";
      return NULL;
  }

  builder.SetInsertPoint(entry_block);
  CodegenSetIsNullArg(codegen, entry_block, false);
  builder.CreateRet(result);

  return codegen->FinalizeFunction(function);
}

}
